src/voxelalgorithms.cpp

```diff
@@ -1187,30 +1276,105 @@ void blit_back_with_light(ServerMap *map, MMVManip *vm,
 
 	vm->blitBackAll(modified_blocks, true);
 
-	// --- STEP 4: Do unlighting
+	// --- STEP 4: Finish light update
 
-	for (size_t bank = 0; bank < 2; bank++) {
-		LightBank b = banks[bank];
-		unspread_light(map, ndef, b, unlight[bank], relight[bank],
-			*modified_blocks);
+	finish_bulk_light_update(map, minblock, maxblock, unlight, relight,
+		modified_blocks);
+}
+
+/*!
+ * Resets the lighting of the given map block to
+ * complete darkness and full sunlight.
+ *
+ * \param light incoming sunlight, light[x][z] is true if there
+ * is sunlight above the map block at the given x-z coordinates.
+ * The array's indices are relative node coordinates in the block.
+ * After the procedure returns, this contains outgoing light at
+ * the bottom of the map block.
+ */
+void fill_with_sunlight(MapBlock *block, INodeDefManager *ndef,
+	bool light[MAP_BLOCKSIZE][MAP_BLOCKSIZE])
+{
+	if (block->isDummy())
+		return;
+	// dummy boolean
+	bool is_valid;
+	// For each column of nodes:
+	for (s16 z = 0; z < MAP_BLOCKSIZE; z++)
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+		// True if the current node has sunlight.
+		bool lig = light[z][x];
+		// For each node, downwards:
+		for (s16 y = MAP_BLOCKSIZE - 1; y >= 0; y--) {
+			MapNode n = block->getNodeNoCheck(x, y, z, &is_valid);
+			// Ignore IGNORE nodes, these are not generated yet.
+			if (n.getContent() == CONTENT_IGNORE)
+				continue;
+			const ContentFeatures &f = ndef->get(n.getContent());
+			if (lig && !f.sunlight_propagates)
+				// Sunlight is stopped.
+				lig = false;
```

Comment:
This is valid but you should put brackets around this block anyway to prevent misunderstandings.